File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-8.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int ans = 0;
    int goodNodes(TreeNode* root) {
        countGood(root, -101);
        return ans;
    }
    void countGood(TreeNode* root, int maxVal) {
        if (!root) {return;}
        if (root->val >= maxVal) {
            ans++;
        }
        maxVal = max(maxVal, root->val);
        countGood(root->left, maxVal);
        countGood(root->right, maxVal);
    }
};

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-8.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        self.i = 0
        pos = dict()
        for i, n in enumerate(inorder):
            pos[n] = i
        def build(l: int, r: int) -> TreeNode:
            if l > r:
                return None
            curr = preorder[self.i]
            root = TreeNode(curr)
            self.i += 1
            idx = pos[curr]
            root.left = build(l, idx - 1)
            root.right = build(idx + 1, r)
            return root
        root = build(0, len(preorder) - 1)
        return root
            